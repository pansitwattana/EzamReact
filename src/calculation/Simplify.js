import math from 'mathjs'
import parser from './Parser'

export default (expression) => {
  if (!expression) {
    console.error('No expression pass through parameter')
    return null
  }
  const expr = parser(expression)
  if (!expr) {
    console.error(`Can not parse ${expression}`)
    return null
  }

  try {
    const result = math.simplify(expr);
    let latex = result.toTex();
    let useToReplace = latex.slice()
    let useToFindFrac = latex.slice()
    let willReplace = ''
    let startIndex = 0
    let endIndex = 0
    let diff = 0
    while (true) {
      const fracIndex = useToFindFrac.indexOf('\\frac')
      if (fracIndex === -1) {
        break;
      }
      startIndex = fracIndex
      endIndex = startIndex + 7
      const splitLatex = latex.slice(fracIndex + 6, latex.length)
      const bracketIndex = splitLatex.indexOf('}')
      const topNumber = splitLatex.slice(0, bracketIndex)
      endIndex += bracketIndex + 1
      if (!isNaN(parseInt(topNumber))) {
        const number = parseInt(topNumber)
        
        const backSplit = splitLatex.slice(bracketIndex + 2, splitLatex.length)
        const bracketIndex2 = backSplit.indexOf('}')
        const botNumber = backSplit.slice(0, bracketIndex2)
        endIndex += bracketIndex2 + 1
        if (!isNaN(parseInt(botNumber))) {
          const number2 = parseInt(botNumber)

          const value = number / number2
          willReplace = latex.slice(startIndex + diff, endIndex)
          useToReplace = useToReplace.replace(willReplace, value.toString())
          useToFindFrac = useToReplace.slice()
          useToFindFrac = useToFindFrac.slice(endIndex, useToFindFrac.length)
          diff += endIndex
          // useToFindFrac = splitLatex.slice(bracketIndex2 + 2, splitLatex.length)
        } else {
          break;
        }
      } else {
        break;
      }
    }
    return useToFindFrac
    // return math.format(result, { precision: 13 });
  } catch (e) {
    console.error(e)
    return null
  }
}