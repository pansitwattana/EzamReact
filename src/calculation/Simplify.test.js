import Simplify from './Simplify'

test('simplify 2 * 1 * x ^ (2 - 1)', () => {
  expect(Simplify('2 * 1 * x ^ (2 - 1)')).toMatchSnapshot()
})

test('simplify 77*3sin(5x/10)', () => {
  expect(Simplify('77*3sin(5x/10)')).toMatchSnapshot()
})

test('simplify 10x+5x+77/3sin(5x/10)', () => {
  expect(Simplify('10x+5x+77/3sin(5x/10)')).toMatchSnapshot()
})