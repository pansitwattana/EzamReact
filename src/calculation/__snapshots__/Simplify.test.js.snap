// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`simplify 2 * 1 * x ^ (2 - 1) 1`] = `"2\\\\cdot x"`;

exports[`simplify 10x+5x+77/3sin(5x/10) 1`] = `"15\\\\cdot x+\\\\frac{\\\\sin\\\\left(\\\\frac{ x}{2}\\\\right)\\\\cdot77}{3}"`;

exports[`simplify 77*3sin(5x/10) 1`] = `"231\\\\cdot\\\\sin\\\\left(\\\\frac{ x}{2}\\\\right)"`;
